_posts/2021-10-29-swift-interface-for-node-js.md

```diff
@@ -0,0 +1,179 @@
+```
+
+Pretty basic. Let’s move on to a simple function that will greet whatever will
+passed as an argument.
+
+My oh-so-naive first draft of interfacable code looked like this:
+
+```swift
+public func greet(name: String) -> String {
+    return "Hello, \(name)!"
+}
+```
+
+But my successful attempt ended up looking like this:
+
+```swift
+@_cdecl("greet")
+public func greet(
+    cStringName: UnsafePointer<Int8>,
+    outputSize: Int16,
+    output: UnsafeMutablePointer<Int8>
+) -> Bool {
+    let name = String(cString: cStringName)
+    let cString = "Hello, \(name)!".cString(using: .utf8)!
+    for (index, char) in cString.enumerated() {
+        if (index >= outputSize) {
+            return false
+        }
+        output.advanced(by: index).pointee = char
+    }
+    return true
+}
+```
+
+Lessons learned:
+
+- An interoperable string is actually called a C String, which translates to a
+  `UnsafePointer<Int8>`;
+- I would get an error returning the output as a result of the function. My
+  guess is that the library would release the memory as soon as the function had
+  returned, hence preventing Node to read something meaningful in memory. So we
+  should assume that the caller will provide us with some writable memory where
+  we will copy our greeting so that Node can read it. And thus, for a proper
+  error management and prevent dreaded `segmentation fault` errors, we need to
+  know how big this memory allocation is;
+- [`nm -gU`][11] is a command line tool that will display the list of external
+  symbols of a dynamic library;
+- In order for the library to include searchable symbols, we need to use [this
+  `@_cdecl` annotation][10] that is still private with Swift.
+
+Rough trip. But it was fun.
+
+## Using the dynamic library from Node
+
+Similarly, this is the end result for the Node code and the lessons learned.
+
+```node
+var ffi = require("ffi-napi");
+
+var myLib = ffi.Library(
+  "../MyLibrary/.build/x86_64-apple-macosx/debug/libMyDynamicLibrary.dylib",
+  {
+    greet: ["bool", ["string", "long", "char *"]],
+  }
+);
+
+const maxStringLength = 20;
+const theStringBuffer = Buffer.alloc(maxStringLength);
+
+const isSuccess = myLib.greet("Mick", maxStringLength, theStringBuffer);
+if (isSuccess) {
+  // Retrieve and convert the result back to a javascript string
+  var theString = theStringBuffer.toString("utf-8");
+  console.log(`The string: ${theString}`);
+  process.exit(0);
+} else {
+  console.error(`Calling greet failed.`);
+  process.exit(1);
+}
+```
+
+- A dependency was required to interoperate with our dylib. Such libs are
+  usually called _FFI_, standing for [_Foreign Function Interface_][5]. The
+  [`ffi`][3] module didn’t work. Not sure why 🤷. [`ffi-napi`][4] did work. Its
+  documentation is good enough to get inspired by examples for simple use cases;
+- The rest was pretty straightforward.
+
+## Epilogue
+
+```shell
+➜  node node index.js
+The string: --Hello, Mick!--
+```
+
+💪🎉
```

Comment:
It's not clear that the article is over, maybe some more text to make that clearer (even though I assume it's in the code).